test/unit/not_initialized_test.lua

```diff
@@ -44,6 +44,5 @@ g.test_insert = function()
     ]])
 
     t.assert_equals(results, nil)
-    t.assert_str_contains(err.err, "Failed for %w+%-0000%-0000%-0000%-000000000000", true)
-    t.assert_str_contains(err.err, "crud isn't initialized on replicaset")
+    t.assert_str_contains(err.err, "Procedure '_crud.fetch_on_storage' is not defined", true)
```

Comment:
If we'll add `call.any()` wrapper for `select(2, next(vshard.router.routeall())):call(<...>)` similar to `call.single()` (but without `bucket_id`) and use it in `reentrant_fetch()`, we'll get more human friendly "crud isn't initialized on replicaset" error as before. AFAIS, all other vshard calls are wrapped.

Reply:
Added a new function `call.any()` and use it in `reentrace_fetch()`. I think my commits are large, so I put patch with new function to a separate commit.

```diff
+++ b/crud/common/call.lua
@@ -145,4 +145,28 @@ function call.single(bucket_id, func_name, func_args, opts)
     return res
 end
 
+function call.any(func_name, func_args, opts)
+    dev_checks('string', '?table', {
+        timeout = '?number',
+    })
+
+    local timeout = opts.timeout or call.DEFAULT_VSHARD_CALL_TIMEOUT
+
+    local replicasets = vshard.router.routeall()
+    local replicaset = select(2, next(replicasets))
+
+    local res, err = replicaset:call(func_name, func_args, {
+        timeout = timeout,
+    })
+    if err ~= nil then
+        return nil, wrap_vshard_err(err, func_name, replicaset.uuid)
+    end
+
+    if res == box.NULL then
+        return nil
+    end
+
+    return res
+end
+
 return call
```

```diff
commit 6b57f84baa2be753d4b1101b8a32d0caf7644670
Author: Sergey Bronnikov <sergeyb@tarantool.org>
Date:   Thu Nov 18 21:58:03 2021 +0300

    upsert() [TO SQUASH]

diff --git a/crud/common/sharding_key.lua b/crud/common/sharding_key.lua
index 11bb7f8..4ad1c43 100644
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -1,7 +1,7 @@
 local fiber = require('fiber')
 local errors = require('errors')
-local vshard = require('vshard')
 
+local call = require('crud.common.call')
 local const = require('crud.common.const')
 local dev_checks = require('crud.common.dev_checks')
 local cache = require('crud.common.sharding_key_cache')
@@ -65,9 +65,7 @@ local sharding_key_fetch_cond = fiber.cond()
 -- metadata by a single one, other fibers will wait for a
 -- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch metadata
 -- if a cache is still empty.
-local function reentrant_fetch(replicaset)
-    dev_checks('table')
-
+local function reentrant_fetch()
     if cache.sharding_key_as_index_obj_map ~= nil then
         return cache.sharding_key_as_index_obj_map
     end
@@ -87,7 +85,7 @@ local function reentrant_fetch(replicaset)
     end
 
     sharding_key_fetch_in_progress = true
-    local metadata_map, err = replicaset:call(FETCH_FUNC_NAME, {}, {
+    local metadata_map, err = call.any(FETCH_FUNC_NAME, {}, {
         timeout = const.FETCH_SHARDING_KEY_TIMEOUT
     })
     if err ~= nil then
@@ -123,10 +121,8 @@ function sharding_key_module.fetch_on_router(space_name)
     dev_checks('string')
 
     if cache.sharding_key_as_index_obj_map == nil then
-        local replicasets = vshard.router.routeall()
-        local replicaset = select(2, next(replicasets))
         local err
-        cache.sharding_key_as_index_obj_map, err = reentrant_fetch(replicaset)
+        cache.sharding_key_as_index_obj_map, err = reentrant_fetch()
         if err ~= nil then
             return nil, err
         end
diff --git a/test/unit/not_initialized_test.lua b/test/unit/not_initialized_test.lua
index 667c10b..60ebcb5 100644
--- a/test/unit/not_initialized_test.lua
+++ b/test/unit/not_initialized_test.lua
@@ -44,5 +44,6 @@ g.test_insert = function()
     ]])
 
     t.assert_equals(results, nil)
-    t.assert_str_contains(err.err, "Procedure '_crud.fetch_on_storage' is not defined", true)
+    t.assert_str_contains(err.err, "Failed for %w+%-0000%-0000%-0000%-000000000000", true)
+    t.assert_str_contains(err.err, "crud isn't initialized on replicaset")
 end
```